google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -181,9 +181,419 @@ class ResumableAsyncStreamingReadWriteRpc {
   virtual future<void> Shutdown() = 0;
 };
 
+template <typename RequestType, typename ResponseType>
+class ResumableAsyncStreamingReadWriteRpcImpl
+    : public ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType> {
+ public:
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      RetryPolicyFactory retry_factory,
+      std::shared_ptr<BackoffPolicy const> backoff_policy, AsyncSleeper sleeper,
+      AsyncStreamFactory<RequestType, ResponseType> stream_factory,
+      StreamInitializer<RequestType, ResponseType> initializer)
+      : retry_factory_(std::move(retry_factory)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        initializer_(std::move(initializer)) {}
+
+  ~ResumableAsyncStreamingReadWriteRpcImpl() override {
+    future<void> shutdown = Shutdown();
+    if (!shutdown.is_ready()) {
+      GCP_LOG(WARNING) << "`Finish` must be called and finished before object "
+                          "goes out of scope.";
+      assert(false);
+    }
+    shutdown.get();
+  }
+
+  ResumableAsyncStreamingReadWriteRpcImpl(
+      ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType>&&) =
+      delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableAsyncStreamingReadWriteRpc<RequestType, ResponseType>&&) =
+      delete;
+
+  future<Status> Start() override {
+    future<Status> status_future;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      status_future = status_promise_.get_future();
+      assert(!retry_promise_.has_value());
+      retry_promise_.emplace();
```

Comment:
I think this is a question of implicit vs explicit nullability, I think the presence of optional<> makes it clear that this may or may not be present and needs to be checked.